aws/resource_aws_api_gateway_rest_api.go

```diff
@@ -51,6 +51,13 @@ func resourceAwsApiGatewayRestApi() *schema.Resource {
 				Type:     schema.TypeString,
 				Computed: true,
 			},
+
+			"endpoint_type": {
```

Comment:
> Nothing is currently calling `d.Set("endpoint_type", ...)` in the read function so Terraform state is not being updated from the API to detect drift.

I can fix that pending discussion...

> It seems this should be nested under an `endpoint_configuration` block and renamed just `type` (or `types` see below) unless there is a good reason to drift from the API.

Yeah, I wasn't sure how to approach it considering the situation (see below).  Plus there's also an impact on the domain name resource.  Right now I think that resource deviates fairly heavily and I kinda continued the trend by having the `EDGE`/`REGIONAL` selection be predicated on whether you provide a regional or "normal" certificate.  I was trying to avoid having multiple points of failure (i.e., you set `type` to `REGIONAL` but provide an edge certificate ARN).  So I get the concern, just not sure of the best way to approach it for both of these resources.  I'm happy to implement whatever you decide.

> The SDK signature is `Types []*string locationName:"types" type:"list"` so it seems like the API supports being both `EDGE` and `REGIONAL`. Is that the case and is there reason to not support it?

I thought the same thing when I was working on this, that it was rather odd that the value is a list when all indications in the UI point to this being an either/or proposition.  Indeed, when I set a domain up manually using the CLI in order to force both `EDGE` and `REGIONAL` I was greeted with this in my console:

> An Edge Optimized Endpoint and a Regional Endpoint are both configured. Please migrate your DNS to either the Regional Domain Name or the Edge Optimized Domain Name and delete the unused endpoint.

So it can be done but I don't think you're supposed to do it.  Maybe they have future plans for it though.

Reply:
Thanks so much for the investigative work. I'll write up a response in the morning.

Reply:
Sorry for the delayed response. I think in this case we should follow the API types and behaviors. It simplifies maintenance of this resource in Terraform and will present more accurate information about the physical state of the resource if multiple endpoints are currently assigned. At worst we can allow the API errors for incorrect configurations to bubble up and create any plan-time validations tuned for those scenarios.

I also believe that allowing the list configuration of types will allow you migrate from one type to another without downtime while you move DNS records, etc.

Reply:
Hi All, I started working on this also, then realized that this PR existed. I also thought it was a bit odd that `types` was a list, however when submitting an update endpoint_configuration request with a blank value, I got this back.
```
* aws_api_gateway_rest_api.test: BadRequestException: Invalid EndpointTypes specified. Valid options are REGIONAL,EDGE,PRIVATE
```
It looks like supporting multiple types might be more relevant with the `PRIVATE` type? 
Thought I'd mention that as I couldn't find it in the docs anywhere but that might give it more sense. I have heard rumours that they're releasing API gateway inside a vpc soon.